k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt

```diff
@@ -0,0 +1,99 @@
+        // Priority: token > password > ssh key
+        when {
+            cred.token.isNotBlank() -> {
+                log.debug("populating token with username in manifest")
+                data.putAll(
+                    mapOf(
+                        "username" to encoder.encodeToString(cred.token.toByteArray()),
+                        "password" to ""
+                    )
+                )
+            }
+            cred.username.isNotBlank() -> {
+                log.debug("populating username and password in manifest")
+                data.putAll(
+                    mapOf(
+                        "username" to encoder.encodeToString(cred.username.toByteArray()),
+                        "password" to encoder.encodeToString(cred.password.toByteArray())
+                    )
+                )
+            }
+            cred.sshPrivateKey.isNotBlank() -> {
+                log.debug("populating private key in manifest")
+                data.putAll(mapOf("identity" to encoder.encodeToString(cred.sshPrivateKey.toByteArray())))
+            }
+        }
+        // setting strategy.spinnaker.io/versioned is needed to avoid creating secrets with versioned names e.g. testing1-v000
+        return K8sObjectManifest(
+            "v1",
+            "Secret",
+            mapOf(
+                "namespace" to resource.spec.namespace,
+                "name" to resource.spec.account,
+                "annotations" to mapOf("strategy.spinnaker.io/versioned" to "false")
+            ),
+            null,
+            data as MutableMap<String, Any>?
+        )
+    }
+
+    override suspend fun current(resource: Resource<CredentialsResourceSpec>): K8sObjectManifest? {
+        log.debug("resource received: $resource spec: ${resource.spec}")
+        try {
+            val res = cloudDriverK8sService.getK8sResource(
+                resource.serviceAccount,
+                resource.spec.locations.account,
+                resource.spec.namespace,
+                "secret ${resource.spec.account}"
+            )
+            log.debug("response from clouddriver: $res manifest: ${res.manifest}")
+            return res.manifest
+        } catch (e: HttpException) {
+            if (e.code() == 404) {
+                logger.info("resource ${resource.id} not found")
+                return null
+            } else {
+                throw e
+            }
+        }
+    }
+
+    override suspend fun getK8sResource(r: Resource<CredentialsResourceSpec>): K8sObjectManifest? {
+        return cloudDriverK8sService.getK8sResource(r)
+    }
+
+    override suspend fun actuationInProgress(resource: Resource<CredentialsResourceSpec>): Boolean =
+        resource.spec.account.let {
+            log.debug(resource.toString())
+            val a = orcaService.getCorrelatedExecutions(it)
+            log.debug("actuation in progress? ${a.isNotEmpty()}: $a")
+            a.isNotEmpty()
```

Comment:
any reason we are not deferring to the parent class?

this is a good check we do here on `resource.spec.account`, that I think we should also do for other `spec.template` elements. but I think we can do it in a different place. If you look at the code for other resources in keel, we can do validation as part of resolving the resource, IIRC. so here, we can remove `actuationInProgress` for the child class all together and defer to the super class. This also relieves us from having to pass orca down all the way to other subclasses too and duplicate the effort. unless we are specifically checking for something here that I am missing?

I'd suggest we look into how to validate and in the super class maybe do a debug log of the entire `spec.template` block to get a feel for the resource as it applies to the cluster.

Reply:
The correlation ID that gets passed to orca is different in `CredentialsResourceHandler` compared to the super class. e.g. the clouddriver account name is passed compared to just the name in `GenericK8sResourceHandler`. I actually think it makes sense to do it in specific handlers. `actuationInProgress()` is implemented in all resource handlers (e.g. `SecurityGroupHandler`)  in Keel as far as I can tell. Without this implemented, excessive amount of `deployManifest` tasks are generated.

Reply:
ok fair. we should implement it for `K8sResourceHandler` and `HelmResourceHandler` too then, right? I am fine if we want to defer that to another PR, but let's track it maybe with an issue so we don't forget.